### PR TITLE
Follower: check if join block different from fetched block

### PR DESCRIPTION
**Pull request by tock-ibm**
Change-Id: I7089df4f1bfc200a5f17582df0da7d4172f8f09b

#### Type of change

- Bug fix

#### Description

Follower: revert check of join block different from fetched block

#### Related issues

Addresses #4264

### CODE REVIEW

**Review by yacovm — CHANGES_REQUESTED** [review]
I don't think that simply removing the check without providing a countermeasure in the same fix is a good idea. 

With this change, the code allows an orderer to pull blocks while disregarding the join block, which obviously was retrieved from somewhere. 

The genesis block cannot be validated, and thus the current code completely puts the validity of the genesis block in the mercy of the orderer it is retrieved from. 

This check lets you know that you have diverged from the chain.

**Comment by tock-ibm** [comment]
> I don't think that simply removing the check without providing a countermeasure in the same fix is a good idea.
> 
> With this change, the code allows an orderer to pull blocks while disregarding the join block, which obviously was retrieved from somewhere.
> 
> The genesis block cannot be validated, and thus the current code completely puts the validity of the genesis block in the mercy of the orderer it is retrieved from.
> 
> This check lets you know that you have diverged from the chain.

@yacovm I agree with what you said. I am changing the mechanism such that:
The follower checks if join block different from fetched block, both header and data, and
- it happens before fetched block is committed, otherwise if the orderer is restarted, the follower continues past that block
- don't panic, since it can affect other channels
- continue to retry, as it might have been an error in a single remote orderer

In the future, we may consider to stop the chain in events like this and reflect it in the status, as proposed in FAB-18106
https://jira.hyperledger.org/browse/FAB-18106

I also added a couple of unit tests for this scenario.

File: orderer/common/follower/follower_chain.go [hunk 1]
```diff
@@ -527,11 +523,24 @@ func (c *Chain) pullUntilTarget(targetHeight uint64, updateEndpoints bool) (uint
...
+				return n, errors.Errorf("block header previous hash mismatch on sequence %d, expected %x, got %x",
 					nextBlock.Header.Number, actualPrevHash, reportedPrevHash)
 			}
+
+			if c.joinBlock != nil && c.joinBlock.Header.Number == nextBlock.Header.Number {
+				if !proto.Equal(nextBlock.Header, c.joinBlock.Header) {
+					c.logger.Errorf("Block header mismatch between the block we pulled and the block we joined with, sequence %d", c.joinBlock.Header.Number)
+					return n, errors.Errorf("block header mismatch between the block we pulled and the block we joined with, sequence %d", c.joinBlock.Header.Number)
+				}
+				if !proto.Equal(nextBlock.Data, c.joinBlock.Data) {
```
Comment (yacovm) [inline orderer/common/follower/follower_chain.go]: why do we need the data check? The block puller assures us the data is the hash in the header, no?

Comment (yacovm) [reply orderer/common/follower/follower_chain.go]: or are you trying to catch a case if we got a join block with a header hash that mismatches its data? if that's the reason then we should check it earlier and not here.

**Comment by yacovm** [comment]
> > I don't think that simply removing the check without providing a countermeasure in the same fix is a good idea.
> > With this change, the code allows an orderer to pull blocks while disregarding the join block, which obviously was retrieved from somewhere.
> > The genesis block cannot be validated, and thus the current code completely puts the validity of the genesis block in the mercy of the orderer it is retrieved from.
> > This check lets you know that you have diverged from the chain.
> 
> @yacovm I agree with what you said. I am changing the mechanism such that: The follower checks if join block different from fetched block, both header and data, and
> 
>     * it happens before fetched block is committed, otherwise if the orderer is restarted, the follower continues past that block
> 
>     * don't panic, since it can affect other channels
> 
>     * continue to retry, as it might have been an error in a single remote orderer
> 
> 

> 
> I also added a couple of unit tests for this scenario.

As long as the follower doesn't evolve into a fully fledged consenter, we're good.

File: orderer/common/follower/follower_chain.go (hunk 1 above)
Comment (tock-ibm) [reply orderer/common/follower/follower_chain.go]: In the fetched block yes, but the join block can be anything... I am not sure we even verify the data hash equals the hash(data) when we accept it via the API... we should do it though

Comment (yacovm) [reply orderer/common/follower/follower_chain.go]: That's what I'm saying  - we should verify the data matches the hash when we digest the block at the time of processing the API call, not **here**.

Comment (tock-ibm) [reply orderer/common/follower/follower_chain.go]: Done

**Review by yacovm — DISMISSED** [review]

**Review by yacovm — APPROVED** [review]

**Pull request merged**